### PR TITLE
Advice provider playground update

### PR DESCRIPTION
**Pull request by frisitano**
This PR achieves the following:
- Updates object names associated with `ProgramInputs` (`operand_stack`, `advice_stack`)
- Adds support for `advice_map` and `merkle_store`
- Refactors input and output validation such that errors are provided via the UI
- Refactors `miden-wasm` backend to return `Result<_, JsValue>` type and never panic
- Updated front end to parse errors returned from the `miden-wasm` backend and display them to the user

TODO:
- update typescript test suite to account for the logic above.  
- Add updates to epxlainer.

Call to action:  It would be great if we could collaborative test the experience to ensure that the tests which have implemented are stabled.

### CODE REVIEW

**Review by Dominik1999 — CHANGES_REQUESTED** [review]

File: playground/miden-wasm/src/lib.rs
Comment (grjte) [inline playground/miden-wasm/src/lib.rs:339]: nit: `proof.proof.unwrap()` is a bit strange. Maybe `proof` above should be renamed to something else (`output`, `result`, etc)?

Comment (grjte) [inline playground/miden-wasm/src/lib.rs:339]: nit: `proof.proof.unwrap()` is a bit strange. Maybe `proof` above should be renamed to something else (`output`, `result`, etc)?

**Comment by Dominik1999** [comment]
@frisitano thanks a lot. That looks great. I left some nits and questions. And in general, I am not sure if we should allow those complex inputs like `merkle_store` in the playground. There is certainly no harm in doing so, but because you cannot create those data structures in the playground, I would expect users to use the CLI anyway for this.

And we need to change all the other examples as well because they don't work now anymore. I can do that after your PR is merged

File: playground/src/utils/helper_functions.tsx
Comment (grjte) [inline playground/src/utils/helper_functions.tsx:181]: does this still work to disable linting for the line with the `any` type? I thought it had to be on the same line (or on the line before it using `eslint-disable-next-line` instead of `eslint-disable-line`)

It would be better to just use an `Error` type instead of `any`, but that doesn't have to happen in this Pr

File: playground/miden-wasm/src/utils_input.rs
Comment (grjte) [inline playground/miden-wasm/src/utils_input.rs:51]: nit: typo in optional

File: playground/miden-wasm/src/lib.rs
Comment (frisitano) [reply playground/miden-wasm/src/lib.rs:306]: I think it makes sense to test that we get a proof which is what we do above with `output.proof.is_some()`. I have then modified the verify function such that the proof that is returned is verified so we have internal consistency.  The persisted proof that we were using before became stale and doesn't work with the latest version of miden-vm.

File: playground/miden-wasm/src/utils_input.rs
Comment (grjte) [inline playground/miden-wasm/src/utils_input.rs:35]: nit: I would use "Merkle" instead of "merkle" everywhere

File: playground/src/components/CodingEnvironment/MidenInputs.tsx
Comment (frisitano) [reply playground/src/components/CodingEnvironment/MidenInputs.tsx:14]: I think 150px limits ones ability to create / edit an advice provider that has significant data input.  Happy to change it back if you think that's best.  Let wait for some feedback from others.

File: examples/advice_provider.masm
Comment (frisitano) [reply examples/advice_provider.masm:2]: The inputs don't really have any logical meaning.  It's just dummy data so we can provide an example of how to access it.  I can change this if needed.  What would you expect to see?

File: playground/src/utils/helper_functions.tsx
Comment (frisitano) [reply playground/src/utils/helper_functions.tsx:66]: The user would have to create the structures manually for both the cli and for the playground. This isn't a great developer experience and I think we should look to build some support for this in the future.  However, merkle data structures are a foundational data structure that form the basis of a lot of applications so I think it's quite important that we support this.  I don't think we should constrain the scope of what's possible in the plagyround.

File: playground/src/utils/helper_functions.tsx
Comment (frisitano) [reply playground/src/utils/helper_functions.tsx:120]: thank you!  I will look into this.

File: playground/src/components/CodingEnvironment/MidenInputs.tsx
Comment (grjte) [inline playground/src/components/CodingEnvironment/MidenInputs.tsx:14]: I think 400px might be too much, since it pushes the large code window quite far down the page on my laptop screen. For most of the simple cases, there isn't that much going on in the inputs, and 400px is more than needed for the outputs. Ideally we would have the option to expand these or keep them smaller, but that's a UI update for the future. I probably wouldn't enlarge it past 250px

File: playground/src/components/CodingEnvironment/MidenOutputs.tsx
Comment (grjte) [inline playground/src/components/CodingEnvironment/MidenOutputs.tsx:78]: Same comment here about the size - maybe 250px instead?

**Review by grjte — COMMENTED** [review]
Thank you! I left a couple of minor non-blocking comments. I'm not able to run this up locally unfortunately (due to network issues, I think), so I can't really judge fully & therefore won't approve at this point.

Regarding my points about the size of inputs/outputs - it's hard for me to really judge this without having been able to try out the new advice inputs, but I think the page looks better when they're smaller, and most of the examples don't use the complex advice inputs

File: playground/src/utils/helper_functions.tsx
Comment (frisitano) [reply playground/src/utils/helper_functions.tsx:25]: yeah it's pretty useful.  It's quite commonly used in association with [`formik`](https://formik.org/docs/guides/typescript) where `yup` provides the input validation.

File: playground/miden-wasm/src/lib.rs
Comment (grjte) [reply playground/miden-wasm/src/lib.rs:306]: I prefer the new approach

File: playground/src/utils/helper_functions.tsx
Comment (frisitano) [reply playground/src/utils/helper_functions.tsx:181]: Thank you for the review!

> And in general, I am not sure if we should allow those complex inputs like `merkle_store` in the playground. There is certainly no harm in doing so, but because you cannot create those data structures in the playground, I would expect users to use the CLI anyway for this.

I think it's quite important that we provide support for merkle data structures as they form the basis of a lot of useful applications as mentioned in my response above.  However, I do agree that we need to build better support for actually building the merke store inputs. 

> And we need to change all the other examples as well because they don't work now anymore. I can do that after your PR is merged

I have a PR lined up for this:
https://github.com/0xPolygonMiden/examples/pull/119/files

File: playground/src/utils/helper_functions.tsx
Comment (grjte) [reply playground/src/utils/helper_functions.tsx:66]: I agree with @frisitano - the CLI and the playground currently have the same problem, so I don't see a reason to limit the functionality in the playground. If people decide they'd rather use the CLI, they can still do so. The problem of finding a better way for creating these structures is something we should tackle in the future for both CLI & playground.

Btw - I think we should have an obvious mention of the CLI/Debugger/REPL somewhere in the playground. Wdyt? Maybe the simplest thing would be to add another link in the navigation that says "CLI/REPL" and links to the user docs page for those? That page is a bit suboptimal though since it doesn't have the github link. Maybe we should update the docs to have instructions to clone the miden-vm repo (or at least provide the link to it). In that case, we would probably want to link to the docs that are deployed from next rather than the ones on wiki.polygon.technology

File: examples/advice_provider.inputs
Comment (Dominik1999) [reply examples/advice_provider.inputs:1]: Nope, let's do that. Was just not sure if you were aware

File: examples/advice_provider.masm
Comment (Dominik1999) [reply examples/advice_provider.masm:2]: When I was reading `Please see the associated `advice_provider.inputs` file for the definition of the advice_provider inputs` I was expecting to get more info in the file about what these data fields are. But that was not in the file. 

We don't really explain what an `advice_map` or a `merkle_store` is. Can we link to the docs in that comment?

**Comment by Dominik1999** [comment]
> Thank you for the review!
> 
> > And in general, I am not sure if we should allow those complex inputs like `merkle_store` in the playground. There is certainly no harm in doing so, but because you cannot create those data structures in the playground, I would expect users to use the CLI anyway for this.
> 
> I think it's quite important that we provide support for merkle data structures as they form the basis of a lot of useful applications as mentioned in my response above. However, I do agree that we need to build better support for actually building the merke store inputs.
> 
> > And we need to change all the other examples as well because they don't work now anymore. I can do that after your PR is merged
> 
> I have a PR lined up for this: https://github.com/0xPolygonMiden/examples/pull/119/files

I am convinced, let's do that.

File: playground/miden-wasm/src/utils_input.rs
Comment (frisitano) [reply playground/miden-wasm/src/utils_input.rs:35]: I'm not sure about this.  I think it would look inconsistent when defining the `.inputs` file.  Here's an example of how it would look.  Happy to change it if you prefer it.

```json
{
    "operand_stack": ["1"],
    "advice_stack": ["1", "2", "3", "4"],
    "advice_map": {
        "0000000000000000000000000000000000000000000000000000000000000000": [5, 6, 7, 8],
        "0000001000000000000000000000000000000000000000000000000000000000": [9, 10, 11, 12]
    },
    "merkle_store": [
        {"MerkleTree": [
            "0000000000000000000000000000000000000000000000000000000000000000", 
            "0000000000000000000000000000000000000000000000000000000000000000", 
            "0000000000000000000000000000000000000000000000000000000000000000", 
            "0000000000000000000000000000000000000000000000000000000000000000"]},
        {"SparseMerkleTree": [
            [1, "0000000000000000000000000000000333000000000000000000000000000000"],
            [3, "0000000000000000000000000000000222000000000000000000000000000000"]
        ]}
    ]
}
```

Comment (frisitano) [reply playground/miden-wasm/src/utils_input.rs:35]: Not sure if you were referring to the comment or the serde renaming.

File: playground/src/components/CodingEnvironment/MidenInputs.tsx
Comment (frisitano) [reply playground/src/components/CodingEnvironment/MidenInputs.tsx]: changed to 250px.

File: playground/miden-wasm/src/lib.rs
Comment (frisitano) [reply playground/miden-wasm/src/lib.rs]: changed

File: playground/miden-wasm/src/utils_input.rs
Comment (frisitano) [reply playground/miden-wasm/src/utils_input.rs]: fixed

File: playground/src/components/CodingEnvironment/MidenInputs.tsx
Comment (frisitano) [reply playground/src/components/CodingEnvironment/MidenInputs.tsx]: changed to 250px

File: playground/miden-wasm/src/utils_input.rs
Comment (grjte) [reply playground/miden-wasm/src/utils_input.rs:35]: Oh yeah, I just meant in the comments. It was a very nitty nit :)

**Pull request merged**
